// vite/dist/buildPluginResolve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const serverPluginHmr_1 = require("./serverPluginHmr");
const vueResolver_1 = require("./vueResolver");
const debug = require('debug')('vite:build:resolve');
exports.createBuildResolvePlugin = (root, cdn, srcRoots, resolver) => {
    return {
        name: 'vite:resolve',
        resolveId(id) {
            if (id === serverPluginHmr_1.hmrClientId) {
                return serverPluginHmr_1.hmrClientId;
            }
            else if (id.startsWith('/')) {
                // if id starts with any of the src root directories, it's a file request
                if (srcRoots.some((root) => id.startsWith(root))) {
                    return;
                }
                const resolved = resolver.requestToFile(id);
                debug(id, `-->`, resolved);
                return resolved;
            }
            else if (id === 'vue') {
                if (!cdn) {
                    return vueResolver_1.resolveVue(root).bundler;
                }
                else {
                    return {
                        id: vueResolver_1.resolveVue(root).cdnLink,
                        external: true
                    };
                }
            }
            else {
                const request = resolver.idToRequest(id);
                if (request) {
                    const resolved = resolver.requestToFile(request);
                    debug(id, `-->`, request, `--> `, resolved);
                    return resolved;
                }
            }
        },
        load(id) {
            if (id === serverPluginHmr_1.hmrClientId) {
                return `export const hot = {}`;
            }
        }
    };
};
